torchsde/_core/methods_strat/diagonal/heun.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from torchsde._core import base_solver
+
+
+class HeunDiagonal(base_solver.GenericSDESolver):
+
+    def step(self, t0, y0, dt):
+        assert dt > 0, 'Underflow in dt {}'.format(dt)
+
+        I_k = [(bm_next - bm_cur).to(y0[0]) for bm_next, bm_cur in zip(self.bm(t0 + dt), self.bm(t0))]
+
+        f_eval = self.sde.f(t0, y0)
+        g_eval = self.sde.g(t0, y0)
+        g_prod_eval = self.sde.g_prod(g_eval, I_k)
+        y1 = [
+            y0_i + f_eval_i * dt + g_prod_eval_i
+            for y0_i, f_eval_i, g_prod_eval_i in zip(y0, f_eval, g_prod_eval)
+        ]
+        
+        t1 = t0 + dt
+        I_k = [(bm_next - bm_cur).to(y1[0]) for bm_next, bm_cur in zip(self.bm(t1 + dt), self.bm(t1))]
```

Comment:
This looks wrong to me. You should only expect to evaluate the Brownian motion once.

Reply:
Search for "Heun" in https://royalsocietypublishing.org/doi/pdf/10.1098/rspa.2003.1247
which might be helpful.

Reply:
Thanks for these materials! Ok, now I see what I got wrong - I needlessly added another `bm` evaluation (which got me better solution without proper conversion) - removed it. So as mentioned the error was due to the lack of conversion between Ito and Strat - For problem `Ex2` I calculated and added correction for `f`:
```python
def f_corr(self, t, y):
    return 0.5 * (-self.p ** 4. * torch.sin(y) * torch.cos(y) ** 5. * (torch.cos(y) ** 2. - 3. * torch.sin(y) ** 2.))
```
And now it solves it way better:
![](https://i.ibb.co/fYjPVvj/15.png)
![](https://i.ibb.co/cNTGvTx/16.png)

And I got such rate:
![](https://i.ibb.co/0F2KXK5/rate.png)

Reply:
And as you mention I will resubmit it to `dev` branch to be up to date with current API design changes

Reply:
Could you share your derivation of the correction? Converting from Ito to Stratonovich, the modified drift is `O(f, g.dg) = O(p^2)` but what you've written is `O(p^4)`.

Also, the rate in Heun (the slope) in your graph is now about 0.1, which is very wrong.

Your implementation now looks correct, though, so I suspect that may be because of the drift being wrong.

Reply:
OK again my error 😅, sorry! Did Strat => Ito instead of the opposite. 

So current rate is (Is it correct?):
![](https://i.ibb.co/JmNdBDc/rate.png)
And all plots look like this:
![](https://i.ibb.co/ncgSRnx/30.png)
And my correction for `Ex2`:
![](https://i.ibb.co/C2FYZjq/IMG-20200816-165546.jpg)

Reply:
Excellent! The rate now looks correct and the Ito/Strat correction matches my derivation.